itest/lndsigner_test.go

```diff
@@ -0,0 +1,840 @@
+	require.NoError(t, err)
+
+	return tctx
+}
+
+// bitcoinCli sends a command to the test context's bitcoind.
+func (tctx *testContext) bitcoinCli(args ...string) map[string]interface{} {
+	tctx.t.Helper()
+
+	bitcoinCliCmd := exec.CommandContext(tctx.ctx, tctx.bitcoincliPath,
+		append([]string{"-datadir=" + tctx.bitcoinDir,
+			"-rpcport=" + tctx.bitcoinRPC, "-rpcuser=user",
+			"-rpcpassword=password", "-rpcwaittimeout=5"},
+			args...)...)
+
+	stdErrBuf := bytes.NewBuffer(make([]byte, 0))
+	bitcoinCliCmd.Stderr = stdErrBuf
+
+	stdOutBuf := bytes.NewBuffer(make([]byte, 0))
+	bitcoinCliCmd.Stdout = stdOutBuf
+
+	err := bitcoinCliCmd.Start()
+	require.NoError(tctx.t, err)
+
+	// If there's an error on exit, show stderr.
+	err = bitcoinCliCmd.Wait()
+	require.NoError(tctx.t, err, string(stdErrBuf.Bytes()))
+
+	stdout := string(stdOutBuf.Bytes())
+
+	// sendtoaddress only returns a txid on success. In this case, the
+	// first argument is "-named".
+	if len(args) > 1 && args[1] == "sendtoaddress" {
+		return map[string]interface{}{
+			"txid": stdout[:64],
+		}
+	}
+
+	// If there's an error parsing the JSON, show stdout to see the issue.
+	resp := make(map[string]interface{})
+	err = json.Unmarshal([]byte(stdout), &resp)
+	require.NoError(tctx.t, err, stdout)
+
+	return resp
+}
+
+// Close cleans up the test context.
+func (tctx *testContext) Close() {
+	tctx.t.Helper()
+
+	for _, lnd := range tctx.lnds {
+		lnd.Close()
+	}
+
+	_ = tctx.bitcoindCmd.Process.Signal(os.Interrupt)
+	_ = tctx.vaultCmd.Process.Signal(os.Interrupt)
+
+	tctx.cancel()
+
+	<-tctx.ctx.Done()
+
+	// os.RemoveAll(tctx.tmpRoot)
+}
+
+// addNode adds a new LND node to the test context, complete with its own
+// lndsignerd. reqPath can be used to specify create or import, reqData must
+// have a network and optional seed/passphrase, and unixSocket may be used to
+// specify that a UNIX socket should be used to communicate between LND and
+// lndsignerd.
+func (tctx *testContext) addNode(reqPath string,
+	reqData map[string]interface{}, unixSocket bool) string {
+
+	tctx.t.Helper()
+
+	resp, err := tctx.vaultClient.Write(reqPath, reqData)
+	require.NoError(tctx.t, err)
+
+	pubKey, ok := resp.Data["node"].(string)
+	require.True(tctx.t, ok)
+	require.Equal(tctx.t, 66, len(pubKey))
+
+	lnd := &lndHarness{
+		tctx:       tctx,
+		idPubKey:   pubKey,
+		unixSocket: unixSocket,
+	}
+
+	lnd.Start()
+
+	tctx.lndMtx.Lock()
+	tctx.lnds = append(tctx.lnds, lnd)
+	tctx.lndMtx.Unlock()
+
+	return pubKey
+}
+
+// waitForSync ensures that each LND has caught up to the blocks that have been
+// mined on bitcoind.
+func (tctx *testContext) waitForSync() {
+	tctx.t.Helper()
+
+	var (
+		blocks, synced int
+		resp           map[string]interface{}
+	)
+
+	for blocks < 1000 {
+		resp = tctx.bitcoinCli("getblockchaininfo")
+		blocks = int(resp["blocks"].(float64))
+	}
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for _, lnd := range tctx.lnds {
+		synced = 0
+
+		for synced != blocks {
```

Comment:
Initially, yes, because the initial generation of 1000 blocks happens in parallel with lnd initialization. This is why the start of the function waits until at least 1000 blocks. But when used directly after generate for all of the other tests, `generate` doesn't return until the blocks are submitted and accepted. In practice, I've never seen it cause a race condition yet. Should I update to expect a block count here?

Reply:
If you really want, you could create a mine function on the context that increases the block height locally. Then in waitforsync wait for that height to be reached?

Reply:
Done: https://github.com/bottlepay/lndsigner/compare/66c3baaad33488e5b25d2e8af0935f3adbec2880..d7ba0804a5348e157f205542fa403bf371d91d80

`mine()` will wait for `getblockchaininfo` to catch up to the block height we requested to mine, so I took that wait out of `waitForSync()` and left it to only check that the `lnd` instances are caught up.